control/setup_control.go

```diff
@@ -0,0 +1,383 @@
+type Options struct {
+	// PositionControlUsingTrapz
+	PositionControlUsingTrapz bool
+
+	// SensorFeedbackVelocityControl
+	SensorFeedbackVelocityControl bool
+
+	// DerivativeType
+	DerivativeType string
+
+	// UseCustomeConfig
+	UseCustomConfig bool
+
+	// CompleteCustomConfig
+	CompleteCustomConfig Config
+
+	// NeedsAutoTuning
+	NeedsAutoTuning bool
+
+	// LoopFrequency
+	LoopFrequency float64
+
+	// ControllableType
+	ControllableType string
+
+	// NeedsSingleAutoTuning
+	NeedsSingleAutoTuning bool
+}
+
+func SetupPIDControlConfig(pidVals []PIDConfig, componentName string, Options Options, c Controllable, logger logging.Logger) (*PIDLoop, error) {
+	pidLoop := &PIDLoop{
+		Controllable: c,
+		logger:       logger,
+		Options:      Options,
+		ControlConf:  Config{},
+		ControlLoop:  nil,
+	}
+
+	// set controlConf as either an optional custom config, or as the defualt control config
+	if Options.UseCustomConfig {
+		pidLoop.ControlConf = Options.CompleteCustomConfig
+		for i, b := range Options.CompleteCustomConfig.Blocks {
+			if b.Type == blockSum {
+				sumIndex = i
+			}
+		}
+	} else {
+		pidLoop.createControlLoopConfig(pidVals, componentName)
+	}
+
+	// auto tune the control loop if needed
+	if Options.NeedsAutoTuning {
+		cancelCtx, cancelFunc := context.WithCancel(context.Background())
+		if err := pidLoop.TunePIDLoop(cancelCtx, cancelFunc); err != nil {
+			return nil, err
+		}
+	}
+
+	if Options.NeedsSingleAutoTuning {
+		pidLoop.StartControlLoop()
+	}
+
+	return pidLoop, nil
+}
+
+func (p *PIDLoop) TunePIDLoop(ctx context.Context, cancelFunc context.CancelFunc) error {
+	var errs error
+	p.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer utils.UncheckedErrorFunc(func() error {
+			cancelFunc()
+			if p.ControlLoop != nil {
+				p.ControlLoop.Stop()
+				p.ControlLoop = nil
+			}
+			return nil
+		})
+		defer p.activeBackgroundWorkers.Done()
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+		// switch sum to depend on the setpoint if position control
+		if p.Options.PositionControlUsingTrapz {
+			p.ControlConf.Blocks[sumIndex].DependsOn[0] = p.BlockNames["constant"][0]
+			if err := p.StartControlLoop(); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+
+			p.ControlLoop.MonitorTuning(ctx)
+		}
+		if p.Options.SensorFeedbackVelocityControl {
+			// to tune linear PID values, angular PI values must be non-zero
+			p.ControlConf.Blocks[angularPIDIndex].Attribute["kP"] = 0.5
+			p.ControlConf.Blocks[angularPIDIndex].Attribute["kI"] = 0.5
+			p.logger.Info("tuning linear PID")
+			if err := p.StartControlLoop(); err != nil {
+				errs = multierr.Combine(errs, err)
+			}
+
+			p.ControlLoop.MonitorTuning(ctx)
+
+			p.ControlLoop.Stop()
+			p.ControlLoop = nil
+
+			// to tune angular PID values, linear PI values must be non-zero
+			p.ControlConf.Blocks[linearPIDIndex].Attribute["kP"] = 0.5
+			p.ControlConf.Blocks[linearPIDIndex].Attribute["kI"] = 0.5
```

Comment:
similar question, though the default linear scaling we use should make this less significant